### ISSUE TITLE
[Feature]: Add ImageNormalizer utility with ImageNet/COCO presets for preprocessing

### ISSUE THREAD

**Issue by GaP040507** [kornia/kornia-rs#791]
### Motivation

Many computer vision inference pipelines require mean/std normalization 
before passing images to a model (e.g., ImageNet or COCO pretrained models).

Currently this preprocessing step must be implemented manually by users. 
Providing a reusable normalization utility inside `kornia-image` could simplify 
common inference pipelines and reduce preprocessing boilerplate.

### Proposed Feature

Introduce an `ImageNormalizer` utility that:

• Supports preset normalization configurations  
  - ImageNet RGB  
  - ImageNet BGR  
  - COCO RGB  
  - COCO BGR  

• Allows custom normalization parameters

• Ensures parameter validation (e.g., std > 0)

• Uses type-safe color format handling through an enum

Example usage:

```rust
use kornia_image::{ImageNormalizer, ColorFormat};

let normalizer = ImageNormalizer::imagenet_rgb();
let normalized = normalizer.normalize(&image)?;

### 📂 Feature Category

Rust Core Library

### 💡 Motivation

Image normalization is a required preprocessing step in many computer vision pipelines.

Typical examples include:

- ImageNet pretrained classification models
- Object detection models trained on COCO
- Vision Transformer models
- Many ONNX/TensorRT inference pipelines

Without a utility, users must manually implement normalization logic for every project, which increases the chance of incorrect preprocessing.

Providing a built-in normalization utility would:

- Reduce preprocessing boilerplate
- Provide reliable presets for common datasets
- Ensure validation of normalization parameters
- Keep preprocessing consistent with Kornia's image abstractions

### 💭 Proposed Solution

Introduce an `ImageNormalizer` struct inside the `kornia-image` crate.

### Key features

• Preset normalization configurations
  - ImageNet RGB
  - ImageNet BGR
  - COCO RGB
  - COCO BGR

• Custom normalization parameters

• Validation of parameters (e.g. `std > 0`)

• Type-safe color format handling using an enum

### Example API

```rust
use kornia_image::{ImageNormalizer, ColorFormat};

let normalizer = ImageNormalizer::imagenet_rgb();
let normalized = normalizer.normalize(&image)?;

### 📚 Library Reference

This feature is based on normalization utilities commonly used in:

- PyTorch torchvision transforms (`Normalize`)
- OpenCV preprocessing pipelines
- ONNX / TensorRT preprocessing workflows

Normalization using dataset-specific mean/std values is a standard preprocessing step used in many computer vision frameworks.

### 🔄 Alternatives Considered

Users could implement normalization manually using tensor operations.

However, this requires repeating the same preprocessing logic across projects and increases the chance of incorrect parameter usage.

Providing a reusable utility inside `kornia-image` simplifies this workflow and provides reliable presets.

### 🎯 Use Cases

Typical use cases include:

• Preprocessing images before inference with pretrained models  
• Preparing input tensors for ONNX / TensorRT pipelines  
• Implementing standardized preprocessing for datasets such as ImageNet or COCO  
• Simplifying preprocessing steps in Rust-based computer vision applications

### 📝 Additional Context

I have implemented a prototype locally with:

- `ImageNormalizer`
- `ColorFormat` enum
- 6 unit tests covering presets, validation, and accessors

All tests pass locally.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

**Comment by Mayank7Bisnewar** [comment]
Hi, I would like to work over this issue.
Please allot it to me..

**Comment by sidd-27** [comment]
theres already an normalize_mean_std function in the code right? @Mayank7Bisnewar

**Comment by Mayank7Bisnewar** [comment]
Yes, that's correct! I saw that the underlying normalize_mean_std function already exists in kornia-imgproc::normalize. However, my goal with ImageNormalizer is to build a higher-level, ergonomic wrapper around it.

While normalize_mean_std handles the core math, developers currently still have to look up the exact mean/std values for datasets like ImageNet or COCO and manually pass them in as arrays. With ImageNormalizer, we can:

Provide Presets: Offer simple constructors like ImageNormalizer::imagenet_rgb() instead of requiring users to hardcode [0.485, 0.456, 0.406] across multiple projects.

Add Validation: Ensure standard deviation values are strictly > 0 to prevent division-by-zero errors at runtime.
Type-safe Color Formats: Use the ColorFormat enum to prevent users from accidentally applying RGB normalizations to BGR images.

Since normalize_mean_std relies on kornia-imgproc (which depends on kornia-image), I'll implement ImageNormalizer inside kornia-imgproc::normalize so it can leverage the existing normalize_mean_std function under the hood without duplicating the tensor math.

**Comment by sidd-27** [comment]
This just feels like it's more annoying to maintain, and would be one more line of code the user has to write.

**Comment by Mayank7Bisnewar** [comment]
That is a very fair point! You're right—since kornia-rs leans heavily toward a clean functional API, forcing users to instantiate an ImageNormalizer struct just to call a method is unnecessary boilerplate.

If the goal is just to save users from looking up these specific float values and making typos, what if we skip the struct entirely and instead just provide standard dataset constants in the normalize module?

**Comment by edgarriba** [comment]
Any agent can look up this constants today. There's no value here and it's too application specific. What it makes sense is fused operators. @sidd-27 did some work around this

**Comment by sidd-27** [comment]
I have a pr which got closed by the bot i think, most of the work was done just some fine-tuning and review were needed I think. If you're upto it feel free to pick it up and try to get that merged
